### PR TITLE
Part 1: Removal of various warnings supression and fixing their warnings, chagning class and delete C++ keywords in names, and changes related to compile common target as C++

### PR DESCRIPTION
**Pull request by hemagx**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

The following warnings suppression has been removed and their warnings were fixed:

- discarded-qualifiers
-  missing-field-initializers
-  incompatible-pointer-types-discards-qualifiers
- deprecated-non-prototype
- unused-but-set-variable
- enum-enum-conversion
- sign-compare

The following keywords has been changed in source:

- `class` -> `class_`
- `delete` -> `delete_`
- `new` -> `new_`

ERS now has ownership over the name given to it, and similarly various functions that were taking `char *` and expecting constant strings were changed to take `const char *` 

**Issues addressed:** 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/clif.c [hunk 1]
```diff
@@ -5921,7 +5921,7 @@ static void clif_skill_cooldown(struct map_session_data *sd, uint16 skill_id, un
 /// Skill attack effect and damage.
 /// 0114 <skill id>.W <src id>.L <dst id>.L <tick>.L <src delay>.L <dst delay>.L <damage>.W <level>.W <div>.W <type>.B (ZC_NOTIFY_SKILL)
 /// 01de <skill id>.W <src id>.L <dst id>.L <tick>.L <src delay>.L <dst delay>.L <damage>.L <level>.W <div>.W <type>.B (ZC_NOTIFY_SKILL2)
-static int clif_skill_damage(struct block_list *src, struct block_list *dst, int64 tick, int sdelay, int ddelay, int64 in_damage, int div, uint16 skill_id, uint16 skill_lv, enum battle_dmg_type type)
+static int clif_skill_damage(struct block_list *src, struct block_list *dst, int64 tick, int sdelay, int ddelay, int64 in_damage, int div, uint16 skill_id, int skill_lv, enum battle_dmg_type type)
```
Comment (MishimaHaruna) [inline src/map/clif.c:5925]: Now that skill_lv is int, there are several places that pass a literal `-1` or `-2` to it, and it might become confusing if at some point in the future a level check is added. Since this is a value that is passed as is to the client, it might be worth defining constants for it if it has a meaning at all. Or at least clearly documented in this function's header that it can take some magic special values.

File: src/map/clif.c [hunk 2]
```diff
@@ -3536,10 +3536,10 @@ static void clif_guild_xy_remove(struct map_session_data *sd)
...
 	nullpo_retv(sd);
 
-	WBUFW(buf,0)=0x1eb;
-	WBUFL(buf,2)=sd->status.account_id;
-	WBUFW(buf,6)=-1;
-	WBUFW(buf,8)=-1;
+	WBUFW(buf,0) = 0x1eb;
+	WBUFL(buf,2) = sd->status.account_id;
+	WBUFW(buf,6) = UINT16_MAX;
+	WBUFW(buf,8) = UINT16_MAX;
```
Comment (MishimaHaruna) [inline src/map/clif.c]: There are several of these, and I see that some have been changed to `UINT16_MAX` while some others have become `0xFFFF`. I'd stick to only one style (possibly `0xFFFF` which makes the actual value clear, since this is a client constant

File: src/map/script.c [hunk 3]
```diff
@@ -25588,7 +25587,7 @@ static BUILDIN(npcskill)
 	struct npc_data *nd;
 	uint16 skill_id             = script_isstringtype(st, 2) ? skill->name2id(script_getstr(st, 2)) : script_getnum(st, 2);
 	unsigned short skill_level  = script_getnum(st, 3);
-	unsigned int stat_point     = script_getnum(st, 4);
+	int stat_point              = script_getnum(st, 4);
```
Comment (MishimaHaruna) [inline src/map/script.c:25629]: After this change, the variable needs to be tested for negative values below in the function

File: src/map/skill.c [hunk 4]
```diff
@@ -4306,14 +4306,14 @@ static int skill_check_condition_mercenary(struct block_list *bl, int skill_id,
 			return 0;
 		}
 
-		int mhp = skill->get_mhp(skill_id, lv);
+		unsigned int mhp = skill->get_mhp(skill_id, lv);
```
Comment (MishimaHaruna) [inline src/map/skill.c]: This function returns signed, so it's probably safe to keep the variable as int and cast within the if below (which does a positive sign check)

File: src/map/skill.c [hunk 5]
```diff
@@ -4306,14 +4306,14 @@ static int skill_check_condition_mercenary(struct block_list *bl, int skill_id,
...
 
 		if (mhp > 0 && get_percentage(st->hp, st->max_hp) > mhp) {
 			clif->skill_fail(sd, skill_id, USESKILL_FAIL_HP_INSUFFICIENT, 0, 0);
 			return 0;
 		}
 
-		int msp = skill->get_msp(skill_id, lv);
+		unsigned int msp = skill->get_msp(skill_id, lv);
```
Comment (MishimaHaruna) [inline src/map/skill.c]: Same as above

File: src/map/skill.c [hunk 6]
```diff
@@ -13122,7 +13122,7 @@ static int skill_castend_pos2(struct block_list *src, int x, int y, uint16 skill
 				// FIXME: Code after this point assumes that the group has one and only one unit, regardless of what the skill_unit_db says.
 				if (ud->skillunit[i]->unit.count != 1)
 					continue;
-				if (distance_xy(x, y, ud->skillunit[i]->unit.data[0].bl.x, ud->skillunit[i]->unit.data[0].bl.y) < r) {
+				if (distance_xy(x, y, ud->skillunit[i]->unit.data[0].bl.x, ud->skillunit[i]->unit.data[0].bl.y) < (unsigned int)r) {
```
Comment (MishimaHaruna) [inline src/map/skill.c]: It's not the case for this skill, but r could be negative (skill unit range accepts -1). Before casting, it might be appropriate to do a >= 0 check.

File: src/map/pc.c [hunk 7]
```diff
@@ -11748,12 +11748,13 @@ static void pc_read_level_penalty_db(void)
...
 static bool pc_read_exp_db_sub_class(struct config_setting_t *t, bool base)
 {
-	struct class_exp_group entry = { { 0 } };
+	struct class_exp_group entry;
 	struct config_setting_t *exp_t = NULL;
 	int maxlv = 0;
 
 	nullpo_retr(false, t);
 
+	memset(&entry, 0x0, sizeof(entry));
```
Comment (MishimaHaruna) [inline src/map/pc.c]: I believe this is going in the opposite direction of the PR's goals. Using `memset()` in c++ is more dangerous, since it'd mess with any non-POD members of a struct.

Instead, this could be something like:

```
struct class_exp_group entry = {
    .name = "",
    .max_level = 0,
    .exp = { 0 },
};
```

(and if the compiler keeps nagging about the `.exp = {0}` part, that can also be expanded).

After switching to c++, this can be {} initialized.

File: src/char/char.c [hunk 8]
```diff
@@ -243,7 +243,7 @@ static void char_set_char_charselect(int account_id)
 		character->pincode_enable = pincode->charselect + pincode->enabled;
 
 	if(character->waiting_disconnect != INVALID_TIMER) {
-		timer->delete(character->waiting_disconnect, chr->waiting_disconnect);
+		timer->delete_(character->waiting_disconnect, chr->waiting_disconnect);
```
Comment (MishimaHaruna) [inline src/char/char.c:246]: Attaching to this unrelated line but I don't believe I have a way to attach a message to a commit. This doesn't do a full replacement of all `delete` variables or members. After this commit I can still see the following:

```
src/api/aclif.c:295:		jsonwriter->delete(sd->json);
src/api/handlers.c:105:			jsonwriter->delete(sd->json);
src/api/handlers.c:122:	jsonwriter->delete(json);
src/api/handlers.c:246:		jsonparser->delete(json);
src/api/handlers.c:254:		jsonparser->delete(json);
src/api/handlers.c:283:	jsonparser->delete(json);
src/api/handlers.c:454:	jsonwriter->delete(json);
src/api/handlers.c:500:	jsonwriter->delete(json);
src/api/handlers.c:530:	jsonwriter->delete(json);
src/api/handlers.c:577:	jsonwriter->delete(json);
src/api/handlers.c:608:	jsonwriter->delete(json);
src/api/jsonparser.c:186:	jsonparser->delete = jsonparser_delete;
src/api/jsonparser.h:56:	void (*delete) (JsonP *ptr);
src/api/jsonwriter.c:120:		jsonwriter->delete(obj);
src/api/jsonwriter.c:134:		jsonwriter->delete(obj);
src/api/jsonwriter.c:172:		jsonwriter->delete(obj);
src/api/jsonwriter.c:185:		jsonwriter->delete(obj);
src/api/jsonwriter.c:208:			jsonwriter->delete(obj);
src/api/jsonwriter.c:286:	jsonwriter->delete = jsonwriter_delete;
src/api/jsonwriter.h:63:	void (*delete) (JsonW *ptr);
src/char/char.c:1933:		inter_homunculus->delete(hom_id);
src/char/char.c:1937:		inter_elemental->delete(elemental_id);
src/char/int_elemental.c:187:	inter_elemental->delete = inter_elemental_delete;
src/char/int_elemental.h:38:	bool (*delete) (int ele_id);
src/char/int_homun.c:274:	inter_homunculus->delete = inter_homunculus_delete;
src/char/int_homun.h:38:	bool (*delete) (int homun_id);
src/char/int_mail.c:454:	inter_mail->delete = inter_mail_delete;
src/char/int_mail.h:44:	bool (*delete) (int char_id, int mail_id);
src/char/int_mercenary.c:231:	inter_mercenary->delete = inter_mercenary_delete;
src/char/int_mercenary.h:43:	bool (*delete) (int merc_id);
src/char/int_quest.c:233:		success &= inter_quest->delete(char_id, old_qd[i].quest_id);
src/char/int_quest.c:263:	inter_quest->delete = inter_quest_delete;
src/char/int_quest.h:33:	bool (*delete) (int char_id, int quest_id);
src/char/mapif.c:425:	bool result = inter_elemental->delete(ele_id);
src/char/mapif.c:937:	bool result = inter_homunculus->delete(homun_id);
src/char/mapif.c:1036:	bool failed = !inter_mail->delete(char_id, mail_id);
src/char/mapif.c:1156:	bool result = inter_mercenary->delete(merc_id);
src/map/atcommand.c:9523:		if (quest->delete(sd, quest_id) != 0) {
src/map/channel.c:443:		channel->delete(chan);
src/map/channel.c:839:		channel->delete(chan);
src/map/channel.c:868:	channel->delete = channel_delete;
src/map/channel.h:109:	void (*delete) (struct channel_data *chan);
src/map/clif.c:19521:		mercenary->delete(sd->md, MERC_DELETE_REMOVED);
src/map/elemental.c:222:	elemental->delete(ed, 0); // Elemental's summon time is over.
src/map/elemental.c:611:	elemental->delete(ed, 1);
src/map/elemental.c:759:			elemental->delete(sd->ed,0);
src/map/elemental.c:1086:	elemental->delete = elemental_delete;
src/map/elemental.h:134:	int (*delete) (struct elemental_data *ed, int reply);
src/map/guild.c:1970:	gstorage->delete(guild_id);
src/map/guild.c:1973:			channel->delete(g->channel);
src/map/guild.c:2503:			channel->delete(g->channel);
src/map/homunculus.c:651:			homun->delete(sd->hd, -1);
src/map/homunculus.c:705:		return homun->delete(sd->hd, E_OMG);
src/map/homunculus.c:742:			return homun->delete(hd, E_OMG);
src/map/homunculus.c:1491:	homun->delete = homunculus_delete;
src/map/homunculus.h:172:	int (*delete) (struct homun_data *hd, int emote);
src/map/instance.c:542:		channel->delete(map->list[m].channel);
src/map/map.c:3755:		channel->delete(map->list[i].channel);
src/map/mercenary.c:263:	mercenary->delete(md, MERC_DELETE_EXPIRED); // Mercenary soldier's duty hour is over.
src/map/mercenary.c:395:	mercenary->delete(md, MERC_DELETE_KILLED);
src/map/mercenary.c:580:	mercenary->delete = merc_delete;
src/map/mercenary.h:139:	void (*delete) (struct mercenary_data *md, int reply);
src/map/pc.c:7957:		mercenary->delete(sd->md, MERC_DELETE_RANAWAY); // Your mercenary soldier ran away.
src/map/pc.c:7960:		elemental->delete(sd->ed, 0);
src/map/pc.c:8979:		elemental->delete(sd->ed, 0);
src/map/quest.c:1175:	quest->delete = quest_delete;
src/map/quest.h:162:	int (*delete) (struct map_session_data *sd, int quest_id);
src/map/script.c:5023:				timer->delete(sd->npc_idle_timer,npc->secure_timeout_timer);
src/map/script.c:23428:			mercenary->delete(md, type);
src/map/script.c:23822:			quest->delete(sd, quest_id);
src/map/script.c:23825:		quest->delete(sd, script_getnum(st, 2));
src/map/skill.c:6873:				elemental->delete(sd->ed,0);
src/map/skill.c:11330:					elemental->delete(sd->ed,0);
src/map/skill.c:11348:					elemental->delete(sd->ed, 0);
src/map/status.c:9303:						mercenary->delete(sd->md, MERC_DELETE_RANAWAY);
src/char/int_storage.c:48:      int delete[MAX_STORAGE] = { 0 };
src/char/int_storage.c:100:                             delete[total_deletes++] = cp_it->id;
src/char/int_storage.c:114:                             StrBuf->Printf(&buf, "%s'%d'", i == 0 ? "" : ", ", delete[i]);
src/map/storage.c:1202:	gstorage->delete = guild_storage_delete;
src/map/storage.h:93:	void (*delete) (int guild_id);
src/map/unit.c:1159:	//timer->delete function does not messes with it. If the function's
```

File: src/map/clif.c (hunk 1 above)
Comment (hemagx) [reply src/map/clif.c:5925]: I will go to document that it can use some magical values, i wanna reduce unrelated changes to this huge PR as much as possible.

File: src/map/script.c [hunk 9]
```diff
@@ -25588,6 +25635,11 @@ static BUILDIN(npcskill)
 
 	nd = map->id2nd(sd->npc_id);
 
+	if (stat_point < 0) {
+		ShowError("npc_skill negative stat point given (%d)\n", stat_point);
```
Comment (MishimaHaruna) [inline src/map/script.c]: Typo: the command is `npcskill` without underscore

File: src/map/packets_struct.h [hunk 10]
```diff
@@ -1870,7 +1870,7 @@ DEFINE_PACKET_HEADER(CZ_CHECKNAME2, 0x0b97)
 struct PACKET_ZC_CHECKNAME {
 	int16 PacketType;
 	int32 CharId;
-	int16 Class;
+	int16 Class_;
```
Comment (MishimaHaruna) [inline src/map/packets_struct.h]: This seems an innocent victim of a case-insensitive search

File: src/map/packets_struct.h [hunk 11]
```diff
@@ -1879,7 +1879,7 @@ DEFINE_PACKET_HEADER(ZC_CHECKNAME, 0x0a51)
...
```
Comment (MishimaHaruna) [inline src/map/packets_struct.h]: This seems an innocent victim of a case-insensitive search

**Review by MishimaHaruna — CHANGES_REQUESTED** [review]

**Review by MishimaHaruna — APPROVED** [review]
